test/fuzz/CMakeLists.txt

```diff
@@ -1,30 +1,109 @@
-        target_link_libraries(fuzzer_${NAME} PRIVATE Boost::url)
-        target_compile_options(fuzzer_${NAME} PRIVATE -g -O2 -fsanitize=fuzzer,address,undefined -fno-sanitize-recover=undefined)
-        target_link_libraries(fuzzer_${NAME} PRIVATE -fsanitize=fuzzer -fuse-ld=lld)
-
-        # Custom target to run fuzzer executable
-        add_custom_target(fuzz_${NAME} fuzzer_${NAME} -rss_limit_mb=8192 -max_total_time=30 -timeout=30 DEPENDS fuzz_${NAME})
-        set_property(TARGET fuzzer_${NAME} PROPERTY FOLDER "fuzzing")
-    endif ()
+# Get number of cores
+include(ProcessorCount)
+ProcessorCount(PROCESSOR_COUNT)
+
+# Determine total fuzz time per file
+file(GLOB BOOST_URL_FUZZER_SOURCE_FILES *.cpp)
+list(LENGTH BOOST_URL_FUZZER_SOURCE_FILES BOOST_URL_FUZZER_SOURCE_FILES_COUNT)
+set(BOOST_URL_FUZZER_TOTAL_TIME_DEFAULT 30)
+math(EXPR BOOST_URL_FUZZER_TOTAL_TIME_DEFAULT "${BOOST_URL_FUZZER_TOTAL_TIME_DEFAULT} / (${PROCESSOR_COUNT} * ${BOOST_URL_FUZZER_SOURCE_FILES_COUNT}) + 1")
+
+# Fuzzing options
+set(BOOST_URL_FUZZER_TOTAL_TIME ${BOOST_URL_FUZZER_TOTAL_TIME_DEFAULT} CACHE STRING "Total time for fuzzing")
+set(BOOST_URL_FUZZER_RSS_LIMIT 8192 CACHE STRING "RSS limit for fuzzing")
+set(BOOST_URL_FUZZER_TIMEOUT 30 CACHE STRING "Timeout for fuzzing")
+set(BOOST_URL_FUZZER_MAX_LEN 4000 CACHE STRING "Maximum size of the input")
+set(BOOST_URL_FUZZER_JOBS ${PROCESSOR_COUNT} CACHE STRING "Number of jobs for fuzzing")
+option(BOOST_URL_FUZZER_ADD_TO_CTEST "Add fuzzing targets to ctest" OFF)
+
+# Corpus
+set(BOOST_URL_FUZZER_CORPUS_PATH ${CMAKE_CURRENT_SOURCE_DIR}/corpus.tar)
+set(BOOST_URL_FUZZER_CORPUS_DIR ${CMAKE_CURRENT_SOURCE_DIR}/corpus)
+if(EXISTS BOOST_URL_FUZZER_CORPUS_PATH)
+    add_custom_target(
+        untar_corpus
+        COMMAND ${CMAKE_COMMAND} -E echo "Untar corpus from ${BOOST_URL_FUZZER_CORPUS_PATH}"
+        COMMAND ${CMAKE_COMMAND} -E tar xvf ${BOOST_URL_FUZZER_CORPUS_PATH}
+        WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}
+        COMMENT "Unzipping corpus"
+        VERBATIM)
+else()
+    add_custom_target(untar_corpus
+        COMMAND ${CMAKE_COMMAND} -E echo "No corpus. Create ${CMAKE_CURRENT_SOURCE_DIR}/corpus."
+        COMMAND ${CMAKE_COMMAND} -E make_directory ${CMAKE_CURRENT_SOURCE_DIR}/corpus
+        COMMENT "Creating corpus directory"
+        VERBATIM)
+endif()
+
+# Target that runs all fuzz targets
+add_custom_target(
+    fuzz
+    COMMAND ${CMAKE_COMMAND} -E echo "Archive corpus from ${BOOST_URL_FUZZER_CORPUS_DIR} to ${BOOST_URL_FUZZER_CORPUS_PATH}"
+    COMMAND ${CMAKE_COMMAND} -E tar cvf ${BOOST_URL_FUZZER_CORPUS_PATH} ${BOOST_URL_FUZZER_CORPUS_DIR}
+    WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}
+    VERBATIM)
+
+# Register a single fuzzer and add as dependency to fuzz target
+function(add_boost_url_fuzzer NAME)
+    # Fuzzer executable
+    set(SOURCE_FILES ${ARGN})
+    add_executable(fuzzer_${NAME} ${SOURCE_FILES})
+    target_link_libraries(fuzzer_${NAME} PRIVATE Boost::url)
```

Comment:
Is it enough to only build the executable with fuzzing flags? Don't you need to build the library with them, too? I use header-only mode for fuzzers so I don't have an answer to the question.

Reply:
Good point. Yes. I think that's "OK" but diagnostics are worse. I'll fix that.